src/spinnaker.cpp

```diff
@@ -0,0 +1,1060 @@
+
+        set_enum_node(camera_->TriggerSelector, genicam_frame_start);
+        set_enum_node(camera_->TriggerSource, to_trigger_source(target.line));
+        set_enum_node(camera_->TriggerActivation,
+                      to_trigger_activation(target.edge));
+        set_enum_node(camera_->TriggerMode,
+                      target.enable ? genicam_on : genicam_off);
+
+        update_input_trigger(last_known_settings_.input_triggers.frame_start);
+    }
+}
+
+void
+SpinnakerCamera::update_output_trigger_exposure(Trigger& trigger)
+{
+    trigger.kind = Signal_Output;
+    trigger.enable = (*(camera_->LineSource) == genicam_exposure_active) &&
+                     (*(camera_->LineSource) == genicam_exposure_active);
+    trigger.line = 1;
+    // TODO: check if this is the expected edge type for exposure active.
+    trigger.edge = TriggerEdge_LevelHigh;
+}
+
+void
+SpinnakerCamera::maybe_set_output_trigger_exposure(Trigger& target)
+{
+    // TODO: is there a way to disable an output line using Spinnaker?
+    if (!is_equal(target, last_known_settings_.output_triggers.exposure) &&
+        target.enable) {
+        set_enum_node(camera_->LineSelector, genicam_line_1);
+        set_enum_node(camera_->LineMode, genicam_output);
+        set_enum_node(camera_->LineSource, genicam_exposure_active);
+        update_output_trigger_exposure(
+          last_known_settings_.output_triggers.exposure);
+    }
+}
+
+void
+SpinnakerCamera::get(struct CameraProperties* properties)
+{
+    const std::scoped_lock lock(lock_);
+    Spinnaker::GenApi::IInteger& binning = get_binning_node(camera_);
+    *properties = {
+        .exposure_time_us = (float)camera_->ExposureTime(),
+        .binning = (uint8_t)binning(),
+        .pixel_type = to_sample_type(*(camera_->PixelFormat)),
+        .offset = {
+          .x = (uint32_t)camera_->OffsetX(),
+          .y = (uint32_t)camera_->OffsetY(),
+        },
+        .shape = {
+          .x = (uint32_t)camera_->Width(),
+          .y = (uint32_t)camera_->Height(),
+        },
+    };
+
+    if (*(camera_->TriggerSelector) == genicam_frame_start) {
+        update_input_trigger(properties->input_triggers.frame_start);
+    }
+
+    if (*(camera_->LineSelector) == genicam_line_1) {
+        update_output_trigger_exposure(properties->output_triggers.exposure);
+    }
+
+    last_known_settings_ = *properties;
+}
+
+void
+SpinnakerCamera::get_meta(struct CameraPropertyMetadata* meta) const
+{
+    const std::scoped_lock lock(lock_);
+    query_exposure_time_capabilites(meta);
+    // Not part of the Spinnaker API.
+    meta->line_interval_us = { .writable = false };
+    meta->readout_direction = { .writable = false };
+    query_binning_capabilites(meta);
+    query_roi_offset_capabilites(meta);
+    query_roi_shape_capabilites(meta);
+    query_pixel_type_capabilites(meta);
+    query_triggering_capabilites(meta);
+}
+
+void
+SpinnakerCamera::query_exposure_time_capabilites(
```

Comment:
```suggestion
SpinnakerCamera::query_exposure_time_capabilities(
```

Reply:
Oops. Thanks for the catch. I fixed these locally so that the declarations and definitions were fixed in one batch.